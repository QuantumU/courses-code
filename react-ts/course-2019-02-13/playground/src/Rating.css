.star-active {
  color: gold;
}

.star-inactive {
  color: lightgrey;
}