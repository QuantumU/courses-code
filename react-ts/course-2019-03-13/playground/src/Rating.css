.star {
  font-size: 3rem;
  color: lightgrey;
  user-select: none;
}

.star.active {
  color: gold;
}
