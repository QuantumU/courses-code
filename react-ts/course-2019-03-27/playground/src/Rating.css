.star-inactive {
  color: lightgrey;
  font-size: 40px;
}

.star-active {
  color: gold;
  font-size: 40px;
}