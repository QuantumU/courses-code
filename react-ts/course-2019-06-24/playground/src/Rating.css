.star {
  font-size: 30px;
  color: lightgrey;
  user-select: none;
}

.star.active {
  color: gold;
}
