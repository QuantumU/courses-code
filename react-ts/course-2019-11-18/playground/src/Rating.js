import React from 'react';

const FULL_STAR = '★';
const EMPTY_STAR = '☆';

function Rating({stars}) {
  const starSpans = [];
  for (let i = 1; i <= 5; i++) {
    starSpans.push(<span key={i}>{i <= stars ? FULL_STAR : EMPTY_STAR}</span>);
  }
  return <div>{starSpans}</div>;
}

function RatingWithMap({stars, onStarsChange}) {
  const starIds = [0, 1, 2, 3, 4, 5];
  if (!starIds.includes(stars - 1)) {
    throw new Error('number of stars must be 1-6');
  }
  const starSpans = starIds.map(i => (
    <span
      key={i}
      onClick={() => {
        if (onStarsChange) {
          onStarsChange(i + 1);
        }
      }}>
      {i + 1 <= stars ? FULL_STAR : EMPTY_STAR}
    </span>
  ));
  return <div>{starSpans}</div>;
}

export default RatingWithMap;
