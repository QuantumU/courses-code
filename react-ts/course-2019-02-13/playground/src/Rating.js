import React, { Component } from 'react';
import './Rating.css';

export default class Rating extends Component {
  /*
    props:
      stars: number
    events:
      onStarsChange: function(number)
  */
  render() {
    const starData = [];
    for (let i = 1; i <= 5; i++) {
      starData.push({
        id: i,
        active: i <= this.props.stars
      });
    }
    return (
      <div>
        {starData.map(star => (
          <span
            key={star.id}
            className={star.active ? 'star-active' : 'star-inactive'}
            onClick={() => {this.props.onStarsChange(star.id)}}
          >
            {star.active ? '★' : '☆'}
          </span>
        ))}
      </div>
    );
  }
}
