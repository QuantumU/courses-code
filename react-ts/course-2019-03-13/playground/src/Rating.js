import React, { Component } from 'react';

import './Rating.css';

class Rating extends Component {
  // props:
  //   stars: number
  // events:
  //   onStarsChange: (stars: number) => void

  shouldComponentUpdate(nextProps, nextState) {
    return this.props.stars !== nextProps.stars;
  }

  render() {
    const starIndexes = [1, 2, 3, 4, 5];

    return (
      <div>
        {starIndexes.map(index =>
          this.props.stars < index ? (
            <span
              className="star"
              key={index}
              onClick={() => {
                this.props.onStarsChange(index);
              }}
            >
              ☆
            </span>
          ) : (
            <span
              className="star active"
              key={index}
              onClick={() => {
                this.props.onStarsChange(index);
              }}
            >
              ★
            </span>
          )
        )}
      </div>
    );
  }
}

export default Rating;
