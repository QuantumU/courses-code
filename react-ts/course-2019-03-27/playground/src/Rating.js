import React from 'react';
import './Rating.css';
import {range} from 'lodash'

const Rating = React.memo(props => {
  /**
   * props:
   *   stars: number
   * event:
   *   onStarsChange: (newStars: number) => undefined
   */
  const starIndexes = range(1, 6);

  return starIndexes.map(index => (
    <span
      className={index <= props.stars ? 'star-active' : 'star-inactive'}
      onClick={() => {
        props.onStarsChange(index);
      }}
      key={index}
    >
      {index <= props.stars ? '★' : '☆'}
    </span>
  ));
});

export default Rating;
