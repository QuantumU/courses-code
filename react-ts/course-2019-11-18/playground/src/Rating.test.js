import React from 'react';
import TestRenderer from 'react-test-renderer';

import Rating from './Rating';

it('renders a rating with 3 stars without crashing', () => {
  const instance = TestRenderer.create(<Rating stars={3} />).root;
});

it('renders a single div', () => {
  const instance = TestRenderer.create(<Rating stars={3} />).root;
  expect(instance.children.length).toBe(1);
  expect(instance.children[0].type).toBe('div');
});

it('renders some spans containing stars', () => {
  const instance = TestRenderer.create(<Rating stars={3} />).root;
  const container = instance.children[0];
  for (let child of container.children) {
    expect(child.type).toBe('span');
  }
});

it('renders 3 active and 3 inactive stars', () => {
  const instance = TestRenderer.create(<Rating stars={3} />).root;
  const container = instance.children[0];
  for (let i = 0; i < 3; i++) {
    expect(container.children[i].children[0]).toBe('â˜…');
  }
  for (let i = 3; i < 6; i++) {
    expect(container.children[i].children[0]).toBe('â˜†');
  }
});

it('triggers an onStarsChange event if a star is clicked', () => {
  const mockFn = jest.fn();
  const instance = TestRenderer.create(
    <Rating stars={3} onStarsChange={mockFn} />,
  ).root;
  const fourthStar = instance.findAllByType('span')[3];
  fourthStar.props.onClick();
  expect(mockFn).toBeCalledWith(4);
});

describe('errors', () => {
  it('throws an error if the number of stars is 0', () => {
    const testFn = () => {
      TestRenderer.create(<Rating stars={0} />);
    };
    expect(testFn).toThrow('number of stars must be 1-6');
  });
  it('throws an error if the number of stars is 7', () => {
    const testFn = () => {
      TestRenderer.create(<Rating stars={7} />);
    };
    expect(testFn).toThrow('number of stars must be 1-6');
  });
});

it('matches snapshot with 2 stars', () => {
  const tree = TestRenderer.create(<Rating stars={2} />).toJSON();
  expect(tree).toMatchSnapshot();
});
