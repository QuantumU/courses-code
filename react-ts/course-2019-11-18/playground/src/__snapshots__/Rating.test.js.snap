// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot with 2 stars 1`] = `
<div>
  <span
    onClick={[Function]}
  >
    ★
  </span>
  <span
    onClick={[Function]}
  >
    ★
  </span>
  <span
    onClick={[Function]}
  >
    ☆
  </span>
  <span
    onClick={[Function]}
  >
    ☆
  </span>
  <span
    onClick={[Function]}
  >
    ☆
  </span>
  <span
    onClick={[Function]}
  >
    ☆
  </span>
</div>
`;
